tests/e2e/watch_test.go

```diff
@@ -311,24 +313,175 @@ func TestDeleteEventDrop_Issue18089(t *testing.T) {
+
+	// revisionChan sends compacted revisions from which watchers should start.
+	revisionChan := make(chan int64)
+	// compactedAfterChanges means that client would compact after 7 put or delete requests.
+	compactedAfterChanges := 7
+
+	// This goroutine will submit changes on $key $totalRev times. It will
+	// perform compaction after every $compactedAfterChanges changes.
+	// Except for first time, the watcher always receives the compacted
+	// revision as start.
+	go func() {
+		defer close(revisionChan)
+
+		lastRevision := int64(1)
+
+		revisionChan <- lastRevision
+		for vi := 1; vi <= totalRev; vi++ {
+			var respHeader *etcdserverpb.ResponseHeader
+
+			if vi%compactedAfterChanges == 0 && performCompactOnTombstone {
+				t.Logf("DELETE key=%s", key)
+
+				resp, derr := c.KV.Delete(ctx, key)
+				require.NoError(t, derr)
+				respHeader = resp.Header
+
+				values = append(values, valueEvent{value: "", typ: mvccpb.DELETE})
+			} else {
+				value := fmt.Sprintf("%d", vi)
+
+				t.Logf("PUT key=%s, val=%s", key, value)
+				resp, perr := c.KV.Put(ctx, key, value)
+				require.NoError(t, perr)
+				respHeader = resp.Header
+
+				values = append(values, valueEvent{value: value, typ: mvccpb.PUT})
+			}
+
+			lastRevision = respHeader.Revision
+
+			if vi%compactedAfterChanges == 0 {
+				revisionChan <- lastRevision
+
+				t.Logf("COMPACT rev=%d", lastRevision)
+				_, err = c.KV.Compact(ctx, lastRevision, clientv3.WithCompactPhysical())
+				require.NoError(t, err)
+			}
+		}
+	}()
+
+	receivedEvents := make([]*clientv3.Event, 0)
+
+	fromCompactedRev := false
+	for fromRev := range revisionChan {
+		watchChan := c.Watch(ctx, key, clientv3.WithRev(fromRev))
+
+		oldEl := len(receivedEvents)
+
+		// firstReceived represents this is first watch response.
+		// Just in case that ETCD sends event one by one.
+		firstReceived := true
+
+		t.Logf("Start to watch key %s starting from revision %d", key, fromRev)
+	watchLoop:
+		for {
+			curEl := len(receivedEvents)
+			if curEl-oldEl == compactedAfterChanges || curEl == totalRev {
+				break
+			}
+
+			select {
+			case watchResp := <-watchChan:
+				t.Logf("Receive the number of events: %d", len(watchResp.Events))
+				for i := range watchResp.Events {
+					ev := watchResp.Events[i]
+
+					// If the $fromRev is the compacted revision,
+					// the first event should be the same as the last event receives in last watch response.
+					if i == 0 && firstReceived && fromCompactedRev {
+						last := receivedEvents[oldEl-1]
+
+						assert.Equal(t, last.Type, ev.Type,
+							"Last received event type %s, but got event type %s", last.Type, ev.Type)
+						assert.Equal(t, string(last.Kv.Key), string(ev.Kv.Key),
+							"Last received event key %s, but got event key %s", string(last.Kv.Key), string(ev.Kv.Key))
+						assert.Equal(t, string(last.Kv.Value), string(ev.Kv.Value),
+							"Last received event value %s, but got event value %s", string(last.Kv.Value), string(ev.Kv.Value))
```

Comment:
nit: Lower case for the first letter of the message

Reply:
Updated.